specs/agents/tracing-spans-limit.md

```diff
@@ -22,34 +22,79 @@ Setting an upper limit will prevent overloading the agent and the APM server wit
 
 ### Implementation
 
+#### Span count
+
+When a span is put in the agent's reporter queue, a counter should be incremented on its transaction, in order to later identify the _expected_ number of spans.
+In this way we can identify data loss, e.g. because events have been dropped.
+
+This counter SHOULD internally be named `reported` and MUST be mapped to `span_count.started` in the intake API.
+The word `started` is a misnomer but needs to be used for backward compatibility.
+The rest of the spec will refer to this field as `span_count.reported`.
+
+When a span is dropped, it is not reported to the APM Server,
+instead another counter is incremented to track the number of spans dropped.
+In this case the above mentioned counter for `reported` spans is not incremented.
+
+```json
+"span_count": {
+  "reported": 500,
+  "dropped": 42
+}
+```
+
+The total number of spans that an agent created within a transaction is equal to `span_count.started + span_count.dropped`. 
+
+#### Checking the limit
+
 Before creating a span,
 agents must determine whether creating that span would exceed the span limit.
-The limit is reached when the total number of created spans minus the number of dropped spans is greater or equals to the max number of spans.
+The limit is reached when the number of reported spans is greater or equal to the max number of spans.
 In other words, the limit is reached if this condition is true:
 
-    span_count.total - span_count.dropped >= transaction_max_spans
-
-The `span_count.total` counter is not part of the intake API,
-but it helps agents to determine whether the limit has been reached.
-It reflects the total amount of started spans within a transaction.
+    span_count.reported >= transaction_max_spans
+
+On span end, agents that support the concurrent creation of spans need to check the condition again.
+That is because any number of spans may be started before any of them end.
+Agents SHOULD guard against race conditions and SHOULD prefer lock-free CAS loops over using locks.
+
+Example with lock:
+```java
+boolean report
+lock()
+report = span_count.reported < transaction_max_spans
+if (report) {
+    span_count.reported++
+}
+unlock()
+```
+
+Example CAS loop:
+```java
+boolean report
+while (true) {
+    int reported = span_count.reported.atomic_get()
+    report = reported < transaction_max_spans
+    if (report && !span_count.reported.compareAndSet(reported, reported + 1)) {
+        // race condition - retry
+        continue
+    }
+    break
+}
+```
```

Comment:
Maybe I'm missing some possible race conditions but I don't think CAS loop is necessary here. If the main challenge is to handle the issue where a span knows if it's eligible to be reported only at its end (for example because eligibility might be determined by span's duration) then we can just keep `eligible_spans_count` (initial value 0)

At the end of each span:
```
    if ( atomic_get( eligible_spans_count ) <= transaction_max_spans && atomic_increment_and_get( eligible_spans_count ) <= transaction_max_spans  ) {
        should_be_reported = true
        atomic_increment( span_count.reported ) // something like java's LongAdder will work better
    } else {
        should_be_reported = false
        atomic_increment( span_count.dropped ) // something like java's LongAdder will work better
    }
```

Optional optimization at the beginning of a span:
```
    might_be_reported = atomic_get( eligible_spans_count ) <= transaction_max_spans
```

Reply:
There can be a race between `atomic_get(span_count.reported) <= transaction_max_spans` and  incrementing the reported count.
Not quite sure what `eligible_spans_count` refers to. What's the difference to `span_count.reported`? Is it another counter in the `span_count` object that's not reported to APM Server? Similar to `reported` but may overshoot it when there are race conditions? I guess that would work and doesn't require CAS loops.

Reply:
> Not quite sure what `eligible_spans_count` refers to. What's the difference to `span_count.reported`? Is it another counter in the `span_count` object that's not reported to APM Server? Similar to `reported` but may overshoot it when there are race conditions? I guess that would work and doesn't require CAS loops.

Yes, exactly.

> There can be a race between `atomic_get(span_count.reported) <= transaction_max_spans` and incrementing the reported count.

I'm not sure I follow.  Since `span_count.reported` is used only for reporting and only `eligible_spans_count` is used to enforce `transaction_max_spans` there should not be any race involving `span_count.reported`, right? As I mentioned above there shouldn't be any need to even do atomic increments on `span_count.reported` and something like java's LongAdder will work better for `span_count.reported` and `span_count.dropped`.

Reply:
Yes, when introducing `eligible_spans_count` there's no possibility for a race condition.